perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_eduPersonScopedAffiliations.java

```diff
@@ -210,28 +207,133 @@ public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDe
+				User user = perunSession.getPerunBl().getUsersManagerBl().getUserById(perunSession, ((AllAttributesRemovedForUser) message).getUser().getId());
+			} catch (AttributeNotExistsException | UserNotExistsException ex) {
+				// silently skip this event, since source attribute couldn't be between deleted
+				// or user no longer exist
+				skip = true;
 			}
-		});
-		handleIdentifiers.add(auditEvent -> {
-			if (auditEvent instanceof AttributeRemovedForUser && ((AttributeRemovedForUser) auditEvent).getAttribute().getFriendlyName().equals(getSecondarySourceAttributeFriendlyName())) {
-				return ((AttributeRemovedForUser) auditEvent).getUser().getId();
-			} else {
-				return null;
+
+			if (!skip) {
+				AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, getDestinationAttributeName());
+				resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AllAttributesRemovedForUser) message).getUser()));
 			}
-		});
-		return handleIdentifiers;
+
+		}
+
+		// handle group attr changes, exclude "members" group, since its not counted within attr value
+
+		if (message instanceof AttributeSetForGroup &&
+				!VosManager.MEMBERS_GROUP.equals(((AttributeSetForGroup) message).getGroup().getName()) &&
+				((AttributeSetForGroup) message).getAttribute().getName().equals(getTertiarySourceAttributeName())) {
+
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, getDestinationAttributeName());
+			// TODO - get only active group users, since expired are not affected by current group affiliations
+			List<User> users = perunSession.getPerunBl().getGroupsManagerBl().getGroupUsers(perunSession, ((AttributeSetForGroup) message).getGroup());
+			for (User user : users) {
+				resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), user));
+			}
+
+		} else if (message instanceof AttributeRemovedForGroup &&
+				!VosManager.MEMBERS_GROUP.equals(((AttributeRemovedForGroup) message).getGroup().getName()) &&
+				((AttributeRemovedForGroup) message).getAttribute().getName().equals(getTertiarySourceAttributeName())) {
+
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, getDestinationAttributeName());
+			// TODO - get only active group users, since expired are not affected by current group affiliations
+			List<User> users = perunSession.getPerunBl().getGroupsManagerBl().getGroupUsers(perunSession, ((AttributeRemovedForGroup) message).getGroup());
+			for (User user : users) {
+				resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), user));
+			}
+
+		} else if (message instanceof AllAttributesRemovedForGroup &&
+				!VosManager.MEMBERS_GROUP.equals(((AllAttributesRemovedForGroup) message).getGroup().getName())) {
```

Comment:
Wrong indentation.

Reply:
File is auto aligned consistently with two tabs, if its break of expression or method call chain. It seems to be like that in other files too. One tab is reserved for default indent, like class and its content, or method and its content or generally any code block and its content.

Reply:
I see it now. Missed that exclamation mark (now i see it is part of if clause).